src/server/tran_server.cpp

```diff
@@ -362,20 +381,47 @@ tran_server::disconnect_all_page_servers ()
 void
 tran_server::disconnect_page_server_async (const connection_handler *conn)
 {
-  std::lock_guard<std::shared_mutex> lk_guard (m_page_server_conn_vec_mtx);
+  bool main_conn_changed = false;
+  std::string prev_main_conn_id;
   auto &conn_vec = m_page_server_conn_vec;
+  auto ulock = std::unique_lock<std::shared_mutex> (m_page_server_conn_vec_mtx);
 
-  for (auto it = conn_vec.begin (); it != conn_vec.end (); it++)
+  auto conn_it = std::find_if (conn_vec.begin (), conn_vec.end (),
+			       [&conn] (std::unique_ptr<connection_handler> &conn_uptr)
+  {
+    return conn_uptr.get () == conn;
+  });
+
+  if (conn_it == conn_vec.end ())
+    {
+      return; // the connection is already cleared in disconnect_all_page_servers()
```

Comment:
Not sure. Maybe this supposition is not true anymore and we should assert this situation.
I can't remember when this "the connection is already cleared in disconnect_all_page_servers()" situation occurs.

Reply:
When the lock is held in `disconnect_all_page_servers()`, but it hasn't cleared the vector yet, the disconnect request can arrive and this function tries to acquire the lock and, in turn, is blocked. When it gets the lock since `disconnect_all_page_servers` releases, the connection_handler has already been removed.